### PR TITLE
[NPU] fix qwen3.5 shared expert dual stream acc

### PR DESCRIPTION
**Pull request by zhaozx-cn**
## Motivation

there is acc issue in qwen3.5 shared expert dual stream, due to apply same shared expert stream for mtp and target model, which could lead to  data corruption.
## Modifications

following deepseek model, use different stream for mtp and target model when enable dual stream.
## Accuracy Tests

Qwen/Qwen3.5-397B-A17B ceval acc
this pr:
92.7
official:
93.0

export ASCEND_USE_FIA=1
export SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=128
export HCCL_BUFFSIZE=2400
export DEEPEP_NORMAL_LONG_SEQ_ROUND=20
export DEEPEP_NORMAL_LONG_SEQ_PER_ROUND_TOKENS=4096

export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
export STREAMS_PER_DEVICE=32
export HCCL_OP_EXPANSION_MODE=AIV
export HCCL_SOCKET_IFNAME=lo
export GLOO_SOCKET_IFNAME=lo

export SGLANG_ENABLE_SPEC_V2=1
export SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1
export SGLANG_NPU_USE_MULTI_STREAM=1
export DEEP_NORMAL_MODE_USE_INT8_QUANT=1

sglang serve \
--model-path  Qwen3.5-397B-A17B-w4a8-mtp \
--attention-backend ascend \
--device npu \
--tp-size 8 \
--chunked-prefill-size 58880 \
--max-prefill-tokens 65536 --max-mamba-cache-size 100 \
--mamba-scheduler-strategy extra_buffer \
--trust-remote-code \
--host 127.0.0.1 --max-running-requests 32 \
--mem-fraction-static 0.65 \
--port 31125 \
--quantization modelslim \
--enable-multimodal --moe-a2a-backend deepep --deepep-mode auto \
--mm-attention-backend ascend_attn  \
--dtype bfloat16 --mamba-ssm-dtype bfloat16 \
--speculative-algorithm NEXTN --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
--speculative-draft-model-quantization unquant
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #31241774475](https://github.com/sgl-project/sglang/actions/runs/31241774475)
Latest PR Test (Extra): :x: [Run #31241774428](https://github.com/sgl-project/sglang/actions/runs/31241774428)

### CODE REVIEW

File: python/sglang/srt/models/qwen2_moe.py [hunk 1]
```diff
@@ -454,9 +451,12 @@ def _forward_deepep(self, hidden_states: torch.Tensor, forward_batch: ForwardBat
             # router_logits: (num_tokens, n_experts)
             router_logits, _ = self.gate(hidden_states)
             if enable_dual_stream:
-                shared_output = shared_expert_on_independent_stream(
```
Comment (silencejade) [inline python/sglang/srt/models/qwen2_moe.py:472]: What defects exist in shared_expert_on_independent_stream?

Comment (zhaozx-cn) [reply python/sglang/srt/models/qwen2_moe.py:472]: shared_expert_on_independent_stream uses a global shared stream for both the MTP and the target model, which causes data corruption.

File: python/sglang/srt/models/qwen2_moe.py [hunk 2]
```diff
@@ -454,9 +451,12 @@ def _forward_deepep(self, hidden_states: torch.Tensor, forward_batch: ForwardBat
...
-                    hidden_states.clone(), self._forward_shared_experts
-                )
+                current_stream = torch.cuda.current_stream()
```
Comment (Hexq0210) [inline python/sglang/srt/models/qwen2_moe.py]: The unique features of npu can be avoided with torch.npu to prevent ambiguity.

File: python/sglang/srt/models/qwen2_moe.py [hunk 3]
```diff
@@ -454,9 +451,12 @@ def _forward_deepep(self, hidden_states: torch.Tensor, forward_batch: ForwardBat
...
+                self.alt_stream.wait_stream(current_stream)
+                with torch.cuda.stream(self.alt_stream):
+                    shared_output = self._forward_shared_experts(hidden_states)
+                    shared_output.record_stream(self.alt_stream)
+                    shared_event = self.alt_stream.record_event()
```
Comment (Hexq0210) [inline python/sglang/srt/models/qwen2_moe.py:474]: This code should be abstracted back into the original cmo.

File: python/sglang/srt/models/qwen2_moe.py [hunk 4]
```diff
@@ -478,7 +478,7 @@ def _forward_deepep(self, hidden_states: torch.Tensor, forward_batch: ForwardBat
             topk_output=topk_output,
         )
         if enable_dual_stream:
-            wait_share_stream()
+            torch.cuda.current_stream().wait_event(shared_event)
```
Comment (Hexq0210) [inline python/sglang/srt/models/qwen2_moe.py]: ditto

Comment (zhaozx-cn) [reply python/sglang/srt/models/qwen2_moe.py]: done

File: python/sglang/srt/models/qwen2_moe.py (hunk 2 above)
Comment (zhaozx-cn) [reply python/sglang/srt/models/qwen2_moe.py]: done

File: python/sglang/srt/models/qwen2_moe.py [hunk 5]
```diff
@@ -478,7 +478,7 @@ def _forward_deepep(self, hidden_states: torch.Tensor, forward_batch: ForwardBat
             topk_output=topk_output,
         )
         if enable_dual_stream:
```
Comment (silencejade) [inline python/sglang/srt/models/qwen2_moe.py]: shared_event may be undefined when dp-attn is enabled.
